### PR TITLE
Throw exc while key doesn't have read accesskey permit.

### PR DESCRIPTION
**Pull request by terencefan**
1. Throw `AzureSignalRAccessKeyNotAuthorizedException` while the given Accesskey doesn't have permission to generate `AccessToken` for clients.
2. Add exception handler for both `ASP.NET Core SignalR` and `ASP.NET SignalR`.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 1]
```diff
@@ -64,46 +67,54 @@ public Task<string> GenerateAadToken()
...
+            await InitializedTask;
+            if (!Authorized)
+            {
+                throw new AzureSignalRAccessTokenNotAuthorizedException();
+            }
             return await base.GenerateAccessToken(audience, claims, lifetime, algorithm);
         }
 
-        public async Task UpdateAccessKeyAsync(IServerNameProvider provider, ILoggerFactory loggerFactory)
+        private async Task UpdateAccessKeyAsyncInner(IServerNameProvider provider, ILoggerFactory loggerFactory)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:78]: naming: append `Core` as convention

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:78]: private method below public ones

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 2]
```diff
@@ -125,7 +136,7 @@ private async Task AuthorizeWithTokenAsync(string endpoint, int? port, string se
                 cancellationToken: token);
         }
 
-        private async Task<bool> HandleHttpResponseAsync(HttpResponseMessage response)
+        private async Task<bool> HandleHttpResponseAsyncInner(HttpResponseMessage response)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:139]: `HandleHttpResponseAsyncCore`

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    /// <summary>
+    /// The exception throws when AccessKey is not authorized.
+    /// </summary>
+    class AzureSignalRAccessTokenNotAuthorizedException : AzureSignalRException
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:9]: add explicit `internal`

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+         /// <summary>
+        /// Initializes a new instance of the <see cref="AzureSignalRAccessTokenTooLongException"/> class.
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:12]: typo AzureSignalRAccessTokenTooLongException

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 5]
```diff
@@ -23,15 +24,17 @@ internal class AadAccessKey : AccessKey, IDisposable
 
         private int initialized = 0;
 
-        private readonly TaskCompletionSource<bool> _authorizeTcs = new TaskCompletionSource<bool>();
+        private readonly TaskCompletionSource<bool> _initializedTcs = new TaskCompletionSource<bool>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:27]: use `new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously)`

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:27]: nit: looks like `<object>` and `TrySetResult(null)` is enough

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 6]
```diff
@@ -74,36 +71,13 @@ public async Task UpdateAccessKeyAsync(IServerNameProvider provider, ILoggerFact
             {
                 return;
             }
+            await Task.Run(() => UpdateAccessKeyAsyncCore(provider, loggerFactory));
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:74]: we don't need to `await` the result? I think for this public method, it is only a trigger, we never expect the caller to await for the result

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 7]
```diff
@@ -137,18 +121,50 @@ private async Task<bool> HandleHttpResponseAsync(HttpResponseMessage response)
...
+                    var source = new CancellationTokenSource(AuthorizeTimeout);
+                    try
+                    {
+                        await AuthorizeAsync(provider.GetName(), source.Token);
+                        Log.SucceedAuthorizeAccessKey(logger, Endpoint);
+                        break;
+                    }
+                    catch (Exception e)
+                    {
+                        _isAuthorized = false;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:154]: should this `false` set move outside of the retry logic, considering the case when the previous auth is succeeded?

Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:154]: If the previous auth is succeeded, we will simply break the for...loop and there will be no retry afterward.

File: src/Microsoft.Azure.SignalR/Utilities/ServiceRouteHelper.cs [hunk 8]
```diff
@@ -55,6 +55,12 @@ internal class ServiceRouteHelper
                 await context.Response.WriteAsync(ex.Message);
                 return;
             }
+            catch (AzureSignalRAccessTokenNotAuthorizedException ex)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/Utilities/ServiceRouteHelper.cs:58]: add ut for these?

File: src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs [hunk 9]
```diff
@@ -173,8 +173,16 @@ private Task ProcessNegotiationRequest(IOwinContext owinContext, HostContext con
             }
 
             var url = provider.GetClientEndpoint(null, originalPath, queryString);
-
-            return GenerateClientAccessTokenAsync(provider, context, url, claims);
+            try
+            {
+                return GenerateClientAccessTokenAsync(provider, context, url, claims);
+            }
+            catch (AzureSignalRAccessTokenNotAuthorizedException e)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/Middleware/NegotiateMiddleware.cs:180]: add ut for it?

**Review by vicancy — APPROVED** [review]

**Pull request merged**